### PR TITLE
Issue 4980 - Prompt maxi

### LINKED ISSUE #4980: Prompt Maxi
### Description

New setting for text-maxi, which will add the ability to write texts with the help of chatbot

### Other plugins or website examples

_No response_

### How to implement it

_No response_

### PR DESCRIPTION
**Pull request by Fybex**
# Description

A new setting was added to text-maxi block: `Prompt`. To use it, firstly, you need to add your Open AI key in maxi dashboard new tab: `Maxi Ai`. In that tab, you also can add more context to make quality of generated text higher.

### IMPORTANT: before testing run `npm ci --legacy-peer-deps`, because new dependency `langchain` was added to communicate with chat models.

Fixes #4980 

# How Has This Been Tested?
The best way to test it is to just click on everything you see, I've written a to-do list below, but to make it simple, just click all the buttons you see in prompt maxi and try different variations.

## Small demo of basic functionality

https://github.com/maxi-blocks/maxi-blocks/assets/46112160/6607fe4c-46eb-44fa-94a9-45102191f162

## Demo with updated prompt maxi editor UI/UX

https://github.com/maxi-blocks/maxi-blocks/assets/46112160/c949c50b-0142-4007-8918-78f5e0bb044a

# Test checklist

**_ Front/Back Testing _**

-   [ ] Check whether all settings in the maxi AI dashboard can be saved
-   [ ] Check if Open AI and Google Maps(I've changed the code a bit) API keys validation in maxi dashboard works as expected
-   [ ] Check if in prompt maxi generate (first) tab all settings are changeable, if they inherit default settings from maxi AI dashboard
-   [ ] Check if text generation after clicking button `Write for me` works
-   [ ] Check if you can modify generated text on modify/history/second tab
-   [ ] Test all modification options
-   [ ] Test if on modify/history/second tab all card button(insert, copy, use settings, delete) work as expected
-   [ ] Test if it's possible to select other cards, if it's simple and easy to switch between them
-   [ ] Test if stop button stops generating text (it appears on generate process, recommend generating body text or use lengthen modification to be able to have time to catch this button, which appears in the bottom of the second tab)
-   [ ] Select text in text-maxi and ensure that in the second tab `Selected text` card appears, which you can select and then modify with modify options
-   [ ] Check if history stores correctly in local storage and doesn't disappear after reload
-   [ ] Test import/export history options
-   [ ] Test if generated/modified text is giving some good and expected in general results. Try to generate/modify text with container/page context and with/without `Website identity` maxi AI dashboard filled to see the difference

**_ Pre-Code Testing _**

-   [ ] Check whether all settings in the maxi AI dashboard can be saved
-   [ ] Check if in prompt maxi generate (first) tab all settings are changeable, if they inherit default settings from maxi AI dashboard
-   [ ] Check if text generation after clicking button `Write for me` works
-   [ ] Check if you can modify generated text on modify/history/second tab
-   [ ] Test all modification options
-   [ ] Test if on modify/history/second tab all card button(insert, copy, use settings, delete) work as expected
-   [ ] Test if it's possible to select other cards, if it's simple and easy to switch between them
-   [ ] Test if stop button stops generating text (it appears on generate process, recommend generating body text or use lengthen modification to be able to have time to catch this button, which appears in the bottom of the second tab)
-   [ ] Select text in text-maxi and ensure that in the second tab `Selected text` card appears, which you can select and then modify with modify options
-   [ ] Check if history stores correctly in local storage and doesn't disappear after reload
-   [ ] Test import/export history options
-   [ ] Test if generated/modified text is giving some good and expected in general results. Try to generate/modify text with container/page context and with/without `Website identity` maxi AI dashboard filled to see the difference
-   [ ] Test the quality and logic of prompts. Check if prompt are well-built and don't show empty inputs from website identify dashboard tab, for example. 
-   [ ] Test if all cases are covered, like error handling, no api key, etc...
-   [ ] Check no commented code and no unnecessary imports
-   [ ] Standards of the project have been followed
-   [ ] No errors/warnings on console

# Checklist

-   [X] My code follows the style guidelines of this project
-   [X] I have performed a self-review of my code
-   [X] I have commented my code, particularly in hard-to-understand areas
-   [X] My changes generate no new warnings/errors
-   [X] New and existing unit tests pass locally with my changes

### Summary by CodeRabbit

Release Notes:

- New Feature: Added a "Maxi Ai" submenu page to the plugin's dashboard, allowing users to access Maxi AI integration settings and generate content.
- New Feature: Introduced a new method `maxi_blocks_maxi_ai()` to generate the content for the "Maxi Ai" tab.
- New Feature: Added support for OpenAI API key validation and error handling.
- New Feature: Implemented a history feature with buttons for importing, exporting, and cleaning history.
- Bug Fix: Improved error handling in the `maxiAdmin` function by logging errors to the console.
- Refactor: Refactored the custom validation logic for Google API key and added a new custom validation logic for OpenAI API key.
- Style: Updated the styling of various elements in the UI, including result cards and modification options.
- Test: Added unit tests for the new components and functionality.

These changes enhance the user experience by providing seamless integration with Maxi AI, improved error handling, and a history feature for managing generated content.

### CODE REVIEW

**Comment by elzadj** [comment]
test image 
screenshot-wordpress local-2023 09 15-17_32_09

**Comment by ckp267** [comment]
Prompt max new styles version "headline" has a recurring (")

image

image

**Review by ckp267 — APPROVED** [review]
@Fybex Please proceed. Lets get this tested and merged asap. Time to get it live into creators hands.

**Comment by Olekrut** [comment]
1) is it should be clickable?
image

**Review by Olekrut — CHANGES_REQUESTED** [review]
1) to add hover colour for remove button
1.1)
image
1.2)
image

2) cropped remove icon for pre-last generation
image
3) is it limit? How it should work, even if set 0 > we get about 40 characters
image
4) Is it ok, that text is adding to begin, not after previous text
https://jam.dev/c/5f50de76-ba6e-4b6d-a7f7-1e24b78fa61a
5)

**Comment by ckp267** [comment]
> 1. to add hover colour for remove button
>    1.1)
>   
>    1.2)
>    > 2. cropped remove icon for pre-last generation
>    
> 3. is it limit? How it should work, even if set 0 > we get about 40 characters
>    
> 4. Is it ok, that text is adding to begin, not after previous text
>   

@Olekrut you can approve **1 & 2** (we will do small style fixes another time - we want to get prompt maxi merged so people can use it)

**3** is ok for now - you can apporve. We added +/- 40 characters because the generation output cannot be exact. You need some flexibility to write the words. 

**4** you can appove - this is correct expected behaviour. To replace - highlight all the text. To "insert" put the cursor. 

**5** nothing to say... lol

**Review by Olekrut — APPROVED** [review]

**Comment by elzadj** [comment]
@Fybex could you please merge the branch ```prompt-maxi-new-styles``` into this one? It adds new styles to the Maxi UI. 

https://github.com/maxi-blocks/maxi-blocks/compare/prompt-maxi...prompt-maxi-new-styles

**Comment by elzadj** [comment]
Skip the styles, just resolve the conflicts, then we can merge

File: core/admin/admin.js [hunk 1]
```diff
@@ -71,23 +89,115 @@ document.addEventListener('DOMContentLoaded', function maxiAdmin() {
...
+						validationMessage =
+							'Invalid API Key, please check your key and try again';
+						break;
+					case 'RefererNotAllowedError':
+						validationMessage =
+							'Referer not allowed, please allow your domain for that key';
+						break;
+					case 'InvalidCharactersError':
+						validationMessage =
+							'Only alphabet, number, "_", "$", ".", "[", and "]" are allowed in the API key.';
```
Comment (elzadj) [inline core/admin/admin.js:128]: Need to add translations to this file. You can do it with something like this:

```
wp_localize_script(
    'maxi-admin',
    'localization',
    array(
        'loading_status_message' => __('Loading status message...', 'your-text-domain'),
        'please_add_api_key' => __('Please add your API key', 'your-text-domain'),
        // ... (other translations)
    )
);
```

**Review by elzadj — CHANGES_REQUESTED** [review]
Tremendous work, thank you. Almost done:
Need to add translations to admin.js, and the menu item to
screenshot-wordpress local-2023 10 05-12_14_07


Also, please open this tab by default (like in the Setting tab):
screenshot-wordpress local-2023 10 05-13_03_28

**Review by elzadj — APPROVED** [review]

**Pull request merged**